activesupport/lib/active_support/message_verifier.rb

```diff
@@ -46,16 +46,17 @@ def initialize(secret, options = {})
     def valid_message?(signed_message)
       return if signed_message.nil? || !signed_message.valid_encoding? || signed_message.blank?
 
-      data, digest = signed_message.split("--")
-      data.present? && digest.present? && ActiveSupport::SecurityUtils.secure_compare(digest, generate_digest(data))
+      parts = signed_message.split('.')
+      parts.size == 3 && parts.all?(&:present?) && untampered?(parts.pop, parts.join('.'))
     end
 
-    # Decodes the signed message using the +MessageVerifier+'s secret.
+    # Decodes the signed message using the +MessageVerifier+'s secret and returns the claims hash.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
     #
-    #   signed_message = verifier.generate 'a private message'
-    #   verifier.verified(signed_message) # => 'a private message'
+    #   signed_message = verifier.generate('a private message', for: 'test', expires_at: 1.day.from_now)
+    #   verifier.verified(signed_message)
+    #   # => {:pld=>"a private message", :for=>"test", :exp=>"2015-06-29T07:16:12.579Z"}
```

Comment:
I am sorry I got didn't think much here :see_no_evil: . I think I know why we are trying to do backward compatibility. You can say that I focused more on getting the output from the whole system rather than thinking about each component and overall scenario.

I will be more careful in future. :purple_heart: :blue_heart: :green_heart: :heart: :sparkling_heart: